lib/gpu/lib/src/formats.dart

```diff
@@ -0,0 +1,28 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+part of flutter_gpu;
+
+/// Specifies where an allocation resides and how it may be used.
+enum StorageMode {
```

Comment:
Does it make sense to expose any sort of StorageMode enum for buffers?

HostBuffers are obviously host visible. DeviceBuffers can't be transient. hostVisible/devicePrivate is the same thing on UMA, and on non-UMA devices I don't know when you'd use hostVisible.

Reply:
This is a great question -- I think for sure we should allow DeviceBuffers to be hostVisible and devicePrivate.

> DeviceBuffers can't be transient.

Might make sense to just have a separate enums `TextureStorageMode` and `BufferStorageMode` TBH.

> hostVisible/devicePrivate is the same thing on UMA either way

Just dumping thoughts here, but while it's true that you get coherency for free either way on UMA (because it's literally the same memory), we're still forced to contend with both architectures in the same interface. So even if `devicePrivate` doesn't technically exist on the backend, we should still throw a validation error when the user is trying to write to stuff they previously marked as `devicePrivate`, lest a developer might release an app that unknowingly fails in a confusing way for some users on some devices.

* devicePrivate resources aren't readable or writable by the CPU at all -- not even for initialization (e.g. `MTLStorageMode.private`).
* hostVisible resources are both readable and writable by the CPU, and the backends can contend with both the unified and separated cases through a unified interface. So: For UMA, use `MTLStorageMode.shared`. For non-UMA, use `MTLStorageMode.managed` and transfer data as necessary. CPU writes get queued immediately. The device itself may be instructed to write to these allocations, and so the device should be the source of truth. Any attempt at _reading_ the data on the CPU should be an async callback in case either a Device->Host transfer is necessary (non-UMA) or a blocking sync is necessary (UMA).

> on non-UMA devices I don't know when you'd use hostVisible

You must use `hostVisible` anytime you're authoring a buffer from host data. OTOH, you might use a `devicePrivate` buffer to store stuff only written to and read by pipelines through storage bindings.

If you don't need reads/writes from the host, you should prefer `devicePrivate` so that the backend/driver can choose the best heap for the job (something which maybe we'll improve over time for some backends), especially on discrete GPUs.